src/System.Private.CoreLib/shared/System/Environment.cs

```diff
@@ -112,7 +113,19 @@ public static string GetFolderPath(SpecialFolder folder, SpecialFolderOption opt
 
         public static bool Is64BitOperatingSystem => Is64BitProcess || Is64BitOperatingSystemWhen32BitProcess;
 
-        public static OperatingSystem OSVersion => s_osVersion.Value;
+        private static OperatingSystem s_osVersion;
+
+        public static OperatingSystem OSVersion
+        {
+            get
+            {
+                if (s_osVersion == null)
+                {
+                    Interlocked.CompareExchange(ref s_osVersion, GetOperatingSystem(), null);
+                }
+                return s_osVersion;
```

Comment:
Make it volatile and use a local to read the static only once when already initialized?

Reply:
We have patterns like this in many places and most of them look like this (no volatile, no caching). It works fine given the guarantees provided by the runtime.

I have been wondering whether it would be a good idea to have a helper (can be internal for now) to let you write this as one-liner without the delegate overhead. Like: 

`return s_osVersion ?? LazyInitializer.Initialize(ref s_osVersion, GetOperatingSystem());`

?

Reply:
Or an overload based on function pointers once they show up in the language?

Reply:
That would make it all `unsafe` ...

Reply:
Yes. But well encapsulated.

Reply:
Anyway, doesn't need to hold this up.

Reply:
I do not think it would be a good idea to encourage use of function pointers and unsafe for something as simple and common as lazy initialization.

Reply:
I'm not sure why it wouldn't be a good idea. As with anything pointer based, you need to understand the ramifications in order to use it, and you get to opt in to doing so. If you're in a situation where the delegate overhead is prohibitive, you can opt in to using the pointer based version, which with simple use would be safe albeit requiring "unsafe". Otherwise I'm not sure an overload like you suggest really adds value; the whole thing can be done with:
```
s_version ??
Interlocked.CompareExchange(ref s_version, GetOS(), null) ??
s_version
``
(plus technically a volatile on the static). Replacing that with:
```
LazyInitializer.EnsureInitialized(ref s_version, &GetOS)
```
(no volatile needed) feels like a meaningful improvement, whereas:
```
s_version ??
LazyInitializer.EnsureInitialized(ref s_version, GetOS())
```
still technically needing volatile (even if coreclr doesn't require it) to me doesn't feel like it moves the needle much. My $.02.